artifactory/utils/buildinfoproperties.go

```diff
@@ -244,8 +250,29 @@ func setProxyIfDefined(config *viper.Viper) error {
 		if err != nil {
 			return errorutils.CheckError(err)
 		}
-		config.Set(PROXY+HOST, host)
-		config.Set(PROXY+PORT, port)
+		if noProxy != "" {
+			//TODO: This is for the Deployer as proxy properties are set only once (clientConfig). Proxy for dependency resolution can be set via the gradle client
+			deployRequest, err := http.NewRequest("HEAD", config.GetString(DEPLOYER_PREFIX+URL), nil)
+			if err != nil {
+				return errorutils.CheckError(err)
+			}
+			deployProxyURL, err := http.ProxyFromEnvironment(deployRequest)
+			log.Debug("Deploy URL:" + config.GetString(DEPLOYER_PREFIX+URL))
+			if err != nil {
+				return errorutils.CheckError(err)
+			}
+			if deployProxyURL != nil {
```

Comment:
There's a chance I'm missing something here, but it looks like ```deployProxyURL``` is used only for the debug logging. My guess is that your intention was to extract the host and the post values from ```deployProxyURL``` before adding them to the properties file.
If this is the case, I suggest simplifying the implementation of this function and doing it once, instead of twice. In other words, we can use the ```http.ProxyFromEnvironment``` API regardless of whether ```noProxy``` is defined or not.

Notice that the tests are also wrong, because when they set no_proxy, the no_proxy value is identical to the http_proxy value, and that's why no host and port are added to the properties file.
The tests should use no_proxy values with wildcards, and check that the no proxy filtering works properly.

Reply:
I agree that this could be simplified. The key element here is the introduction of the `http.ProxyFromEnvironment` API which introduces a robust “NoProxy” logic in that if it returns  nil, nil then you should not set the proxy flag and if it returns nil, <proxy_url> then you should set the proxy flag to that URL. 

As you suggest I could effectively pull most of this code out (I tried to keep the flow of the original code) and replace it with http.ProxyFromEnvironment. I think it also was this way due to the fact that I originally thought that it mattered if we set it for Deploy vs Resolve but after digging in I realized that Gradle handles that part itself.

Reply:
As for the tests there are two sets of tests. One set where the No Proxy is identical and one where they differ.

Reply:
We could add tests that test the full gambit of no_proxy rules, I figured that was best left to the `http.ProxyFromEnvironment` team/codebase instead I simply wanted a test that proved if the proxy was set or not depending on the variables we passed